Update of minikube-ingress-dns to dns2 and new maintainer

cirix opened the pull request:
# Fix for #17872 
## Ingress-dns plugin has been heavily used by developer community but it has not see any maintenance the last 3-5 years. The tool is based on a DNS implementation called dns2 available on npmjs.

1. I have upgraded the version of dns2 and adjusted the code to work as there were breaking changes between the old version(1.3.0) to the new version (2.1.0)
2. The project is hosted on github and also on my open dockerhub where I host the image.
3. The image created has been tested with existing version on minikube.
4. The project in github builds at the moment amd64 and arm64 native images
5. Many security concerns have been addressed and code + container is available for everyone.

Before:
`minikube addons enable ingress-dns`
After:
`minikube addons enable ingress-dns`

Effectively we enable the addon with native linux amd64/arm64 support so that it can run on linux/macosx.

I have validated the build also in my macbook pro running macosx sonoma with M2 CPU

![Screenshot 2024-01-10 at 21 38 19](https://github.com/kubernetes/minikube/assets/924122/516837cf-f53b-40a6-a3e8-e669a7a9b8ad)


![Screenshot 2024-01-10 at 21 34 29](https://github.com/kubernetes/minikube/assets/924122/057f3c37-9911-40d3-8be8-1a08e8c37e47)

@cirix (Nikolaos Ballas)

medyagh reviewed (commented):
Thank you very much for this PR do you mind putting Before After This PR output ? or an example of usage,

also please consider adding your Github username as the new maintainer of this addon

medyagh commented:
btw do you mind refreencing an issue with details of what was the issue you were trying to fix and also link to the github of the source code

cirix commented:
why the pipeline failed?Those are packages already existing, the only change is the entry on addons.go with the replacement of the ingress-dns addon to point the newly build repo

cirix commented:
Ok,
I found my mistake, I have removed due to my error to deps on addons.go. Fixed build is passing in my fork. Effectively I am building arm64/amd64 images and I have upgraded the addon to the latest stable. I can't  find though where in docs I should put my name as a maintainer. Also is it ok that the pull happens against a dockerhub registry and my account?The repository is public and I run snyk on the image. Just checking if you would like to store the image in a different registry.

br
Nik

spowelljr commented:
@cirix The part of the code you modified is only used if the user is using a Kubernetes version less than 1.19.0 which is very rare. If you wanted to update the addon for the average use case you want to replace the image in `pkg/minikube/assets/addons.go`

cirix commented:
@spowelljr  thanks, will adapt the change there to. I adjusted the fix, will create a new pr after my build finishes

![Screenshot 2024-01-10 at 22 13 28](https://github.com/kubernetes/minikube/assets/924122/51a9b5c0-98d9-4476-a486-855061cbd87f)

cirix commented:
> @cirix The part of the code you modified is only used if the user is using a Kubernetes version less than 1.19.0 which is very rare. If you wanted to update the addon for the average use case you want to replace the image in `pkg/minikube/assets/addons.go`

looks more promising
![Screenshot 2024-01-10 at 22 32 26](https://github.com/kubernetes/minikube/assets/924122/4da48ceb-ad4a-4e80-9f89-91f103751b5c)

spowelljr commented:
I would recommend reverting the change to the legacy ingress image (cryptexlabs) as that's used for compatibility issues for Kubernetes versions before 1.19.0. Unless your image also works with the older versions of Kubernetes and you can demonstrate that.

cirix commented:
Yes I have tested it with older versions is a maintenance upgrade fixing also some security concerns . I can provide evidence in form of what again screen shots. Above the discussion already took place changes are applied for both old and new versions .

cirix commented:
@spowelljr , I plan supporting officially arm64 which currently the cryptelex doesn't offer one and dns nmpjs version they have is 5 years old, with some interesting security issues. I have upgraded all libraries to latest stable and ported it to latest stable version of dns2. Please find bellow the image of my tests with version 1.19.0 of kubernetes on the build of minikube on my local machine. I am testing also the compilation on my Macbook Pro with M2 as seen above. Is arm64 something of interest for the modules as compatibility or not? Let me know your thoughts.

![Screenshot 2024-01-13 at 12 51 30](https://github.com/kubernetes/minikube/assets/924122/582f75e2-31a1-42aa-ae9b-7bf9950e0f31)
![Screenshot 2024-01-13 at 12 56 21](https://github.com/kubernetes/minikube/assets/924122/3c0c499c-5242-4898-afcd-14aa8f870907)

cirix commented:
I have validated the change and image with different flavors of k8s from 1.19 -> 1.26 . Also on my fork the builds are passing successfully. Let me know if it;s ok to allow testing.

cirix commented:
Hi all, @minikube-bot  @medyagh  @spowelljr  shall I close this PR. Eventhough I think is the right move forward to update the version of the component. Please let me know your decision. The build and functionality has been validated also for backwards compatibility

dotnich-io commented:
Is there a good reason not to move forward with this?

spowelljr commented on pkg/minikube/assets/addons.go:
```diff
@@ -554,9 +554,9 @@ var Addons = map[string]*Addon{
 			"ingress-dns-pod.yaml",
 			"0640"),
 	}, false, "ingress-dns", "minikube", "", "https://minikube.sigs.k8s.io/docs/handbook/addons/ingress-dns/", map[string]string{
```
```suggestion
	}, false, "ingress-dns", "3rd party (cirix)", "cirix", "https://minikube.sigs.k8s.io/docs/handbook/addons/ingress-dns/", map[string]string{
```

spowelljr reviewed (changes requested):
Since minikube no longer owns the image update the maintainer info

spowelljr commented:
/ok-to-test

spowelljr commented:
The ingress test seems to be consistently failing on all environments

medyagh commented:
@cirix do you mind pulling upstream and making sure the tests pass ?

cirix commented:
Yes will merge upstream and rerun on my fork

cirix commented:
So the podman build fails for me, with an error on major/minor numbers mix, most likely architecture definitions fail. See image attached. All other builds succeded

![Screenshot 2024-04-23 at 12 02 19](https://github.com/kubernetes/minikube/assets/924122/57305f0f-4d7f-4315-9496-bb13ef545fc8)
Actions run on my fork
![Screenshot 2024-04-23 at 12 03 35](https://github.com/kubernetes/minikube/assets/924122/36802cfd-e447-4144-a7f5-4ea6d413ee00)

@medyagh

spowelljr commented:
The Podman functional tests are broken so you can ignore them, but the integration tests are failing on `TestAddons/parallel/Ingress` as shown in https://github.com/kubernetes/minikube/pull/17926#issuecomment-2072350536

medyagh commented:
@cirix please ping me on slack once the tests are passing, so I could take another look at it

spowelljr commented:
Hi @cirix, I tried your PR to figure out the problem, it looks like you might have built the binary in your image for arm64, but pushing it to amd64 images as well. On amd64 I'm getting `exec /usr/local/bin/docker-entrypoint.sh: exec format error`, if you resolve that I think you'll get your tests passing
